test/metadata_test.go

```diff
@@ -94,21 +121,21 @@ func (s) TestInvalidMetadata(t *testing.T) {
 
 	for _, test := range tests {
 		ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
-		defer cancel()
-
 		ctx = metadata.NewOutgoingContext(ctx, test.md)
+		ctx = metadata.AppendToOutgoingContext(ctx, test.appendMD...)
 		if _, err := ss.Client.EmptyCall(ctx, &testpb.Empty{}); !reflect.DeepEqual(test.want, err) {
 			t.Errorf("call ss.Client.EmptyCall() validate metadata which is %v got err :%v, want err :%v", test.md, err, test.want)
 		}
+		cancel()
```

Comment:
Instead of moving these `cancel` calls, could you please do something like:

```go
for _, test := range tests {
	t.Run("unary " + test.name /* define per test-case */, func() {
		ctx, cancel := ...
		defer cancel()
		...
	})
}

// and ... "streaming " + test.name
```